lib/poseidon/sync_producer.rb

```diff
@@ -71,7 +72,8 @@ def send_messages(messages)
       end
 
       if messages_to_send.pending_messages?
-        raise "Poseidon: Failed to send all messages: #{messages_to_send.messages} remaining"
+        failed_messages = messages_to_send.messages
+        raise Errors::MessageFailedToSendError.new(failed_messages, "Poseidon: Failed to send #{failed_messages.size} out of #{total_message_count} messages")
```

Comment:
Good catch. Indeed, bad things happen when an exception is thrown there. I've addressed this by catching and ignoring the `UnableToFetchMetadata` exception in the one `refresh_metadata` call in the send attempts loop.

Does that seem like enough defense to you? I don't expect to get many errors otherwise—and this will avoid masking real bugs. Also, it looks like it better matches the spirit of the surrounding code. (If we can't fetch metadata, keep retrying. It could be a minor network blip; there's no reason to bail out early.)